### PR TITLE
task-list directive

### PR DESCRIPTION
**Pull request by negue**
This taskList-directive changes the classes of the ion-item & button icon and calls the right score-method.

And all this without adding any watches.

With this change we save atleast 2 watches per item.

Note: I readded the one time bindings, because the directive takes care of it

### CODE REVIEW

**Comment by lefnire** [comment]
Checked out this branch, a few things off the bat (uploading a gif to demonstrate at the end of this comment):

1. `+` or `-` on a habit causes no notifications, no changes, and `TypeError: Cannot read property 'enabled' of undefined` (full error [here](https://gist.github.com/lefnire/2842a5acf6386d26886d))
1. `√` a daily does nothing to the daily/stats, but instead brings user to edit screen.

Those are the bugs, I have more to say about this in general in follow-up comment. Generally love the approach, but I wanna bounce off other devs

mobile perf

**Comment by negue** [comment]
oh... I forgot a file... one moment

**Comment by lefnire** [comment]
@Alys @paglias @crookedneighbor @SabreCat (and others I may be missing):

`taskClasses()` is one of the biggest (very likely *the* biggest) bottleneck in front-end perf, both mobile & web. The reason is in Angular if you have a function watcher `taskClasses()` rather than watchers of specific variables (`ng-if='user.stats.hp && user.items.gear.whatever &&...`), then that function watcher will get called every digest cycle, rather than smartly watching for individual changes. As most y'all can attest when debugging tickets on GH, when users complain about perf the first question is "how many tasks do you have?" and it's because of `taskClasses()`.

Ok, so that's the problem. I've tried refactoring the function / removing it entirely, to come up with a variable-watching scheme rather than a function; but no luck. So what @negue is doing here is creating a directive that watches for interactions with tasks to update its colors / check-status / etc, which is super smart. The downside is that it's more complex / less intuitive code (a maintainability concern). IMO it's a worthwhile trade off for better perf, but I wanted to bounce off y'all to see if anyone had other suggestions? If we merge this in, I'd like to put it into habitrpg-shared (or "common", wink wink @crookedneighbor) or something - utilizing the `window.env.IS_MOBILE` variable to handle difference in classes / frameworks.

One big pitfall I can see @negue is that changes won't pick up from 3rd-party interactions until you've interacted with a local task. Eg, if you check off a task on web, that task won't show as checked off on mobile - even after swiping to sync (am I correct?).

**Comment by negue** [comment]
what do you think is it okay to include all scripts with a wildcard? 

for the habits I'll look if it happens to me too

**Comment by lefnire** [comment]
Should be ok to include scripts/**/*. IIRC I've removed any .bk / .tmp js
files we had in there

**Comment by negue** [comment]
and yes the current code won't update the classes on the sync , I think the directive should listen to an event for this

**Comment by lefnire** [comment]
https://github.com/HabitRPG/habitrpg-shared/blob/develop/script/userServices.js#L112. What's it, `$rootScope.$on('userSynced', function(){})` or something?

**Comment by negue** [comment]
yep, this should do it, I'll try that later / tmrw

**Comment by crookedneighbor** [comment]
I like where this is heading!

**Comment by SabreCat** [comment]
Very promising. Do we have any sense of how much improvement we'll see, percentage-wise?

**Comment by lefnire** [comment]
Shot in the dark, it's 60-70% of the perf bottleneck on the tasks page?

File: views/app/list.jade [hunk 1]
```diff
@@ -1,5 +1,5 @@
 mixin taskText
-  markdown.task-text(text=' ::task.text', remove-watch='true')
+  markdown.task-text(text=' ::task.text', remove-watch='true', use-timeout=':: $index > 10', timeout-time=':: ($index > 10) ? $index * 7 : 50')
```
Comment (negue) [inline views/app/list.jade]: Notes: On the first ten tasks the markdown with be parsed immediately, after that all other "invisible" tasks get to be parsed in an increasing timeout , which speeds up the list generation

**Comment by negue** [comment]
The latest checking needs HabitRPG/habitrpg-shared#441

**Comment by lefnire** [comment]
latest changes, tasks still aren't updating color on interaction. Eg, create a new habit, + it a few times - should turn green, but doesn't. Go off the page and come back, it's green

**Comment by negue** [comment]
that is strange ...I'll test it again later

**Comment by negue** [comment]
You were right, seems that I had completely ignored the habits tab.

Now this is working too^^

**Comment by crookedneighbor** [comment]
Woohoo!

File: gulpfile.js [hunk 2]
```diff
@@ -41,23 +41,7 @@ var paths = {
...
-    'scripts/controllers/chatCtrl.js',
-    'scripts/controllers/inventoryCtrl.js',
-    'scripts/controllers/petDetailsCtrl.js',
-    'scripts/controllers/mountDetailsCtrl.js',
-    'scripts/controllers/stableCtrl.js',
-    'scripts/controllers/partyCtrl.js',
-    'scripts/controllers/notificationCtrl.js',
-    'scripts/controllers/guildCtrl.js',
-    'scripts/controllers/purchaseCtrl.js'
+    'scripts/**/*.js'
```
Comment (negue) [inline gulpfile.js]: this change wasn't merged?

Comment (lefnire) [inline gulpfile.js]: Hmm... i just did `git merge negue/taskListDirective, i'll just add it in manually

Comment (lefnire) [inline gulpfile.js]: So strange, that shouldn't have happened. Anyway, https://github.com/HabitRPG/habitrpg-mobile/commit/74a28abd197cd1d8f990129c46f77114b326dc21 . Everything else look ok?

Comment (negue) [inline gulpfile.js]: Seems so^^, thanks!

**Pull request merged**